[ Albert ] Step3-DB에 저장하기

AlbertImKr opened the pull request:
## 안녕하세요 . Jane🤩🤩🤩🤩
![image](https://ssl.pstatic.net/sstatic/keypage/outside/scui/greeting_card/img/180509_card/hello_2_3.png)

구현내용🧑‍💻
---
- Feedback 적용 했습니다.
   - 주말에 한번 정리하려고 합니다. 
- Repository를 DB로 연결 했습니다.
- Exception 처리했습니다.
- Amazon 배포했습니다.

고민내용🧐
---
- 언제, 누구 , 왜 에러를 던져야 하는 가에 대하여 고민을 많이  했습니다.
  - 잘못된 URL에 대하여 Error페이지와 매핑 하도록 예외 처리를 했습니다.
  - 사용자의 입력오류로 인한 오류는 Optional를 사용하여 재 입력 처리를 했습니다. 대신 로그를 남기도록 했습니다.
  - 자료도 많이 찾아봤는데 아직 잘 모르겠어요.
  
아쉬운 점😞
---
- Spirng JDBC로만 DB 구성해야 해요.
- 다음 미션에 Session이 있어서 이번 미션에서 Users와 Posts를 Foreign key로 묶지 못했어요.
- 화면이 너무 안 예쁘요.(무조건 개선할 예정)

다음주 계획⏰
---
- Spring JPA로 변환 해보려고 합니다.
- 조회 Query는 Query DSL 사용 해보려고 합니다.
- Users와 Posts를 관계를 묶으려고 합니다.


URL Convention🗺️
---
| url                    | method | 기능                                |
|------------------------|--------|-----------------------------------|
| /users/login           | get    | 로그인 페이지 요청                        |
| /users/login           | post   | 로그인 요청                            |
| /users/join            | get    | 회원 가입 페이지 요청                      |
| /users                 | post   | 회원 가입 요청                          |
| /users/{userId}        | get    | 회원 프로필 페이지 요청                     |
| /users/{userId}/update | get    | 회원 프로필 설정 페이지 요청                  |
| /users/{userId}/update | put    | 회월 프로필 설정 요청                      |
| /users                 | get    | 모든 회원 맴버 페이지 요청                   |
| /posts/new,/posts/form | get    | 게시글 작성 페이지 요청                     |
| /posts                 | post   | 게시글 추가 요청                         |
| /                      | get    | 모든 게스글 표시되는 메인 페이지 요청             |                    
| /posts/{postId}        | get    | 게시글의 페이지 요청                       |

AlbertImKr committed:
docs: README에 Commit Log Guidelines 기준을 추가

AlbertImKr committed:
refactor: UsersForm 네이밍을 UserForm으로 변경

복수보다 단수를 사용하는것이 더 맞다고 이해 됐다.

AlbertImKr committed:
feat: User에 Builder 추가

AlbertImKr committed:
refactor: User new 생성을 Builder로 변경

AlbertImKr committed:
feat: ProfileForm에 Builder를 구현

AlbertImKr committed:
refactor: User의 ProfileForm mappint 기능을 ProfileForm.Builder으로 수정

AlbertImKr committed:
feat: ProfileSettingForm에 Builder 구현

AlbertImKr committed:
refactor: User의 setter를 제거

AlbertImKr committed:
feat: Post에 Builder 구현

AlbertImKr committed:
feat: Post에 Builder 구현

AlbertImKr committed:
refactor: Post에 Builder를 사용해 new Post를 대체

AlbertImKr committed:
feat: SimplePostForm에 Builder 구현

AlbertImKr committed:
refactor: SimplePostForm에 builder을 사용하여 new 를 대체

AlbertImKr committed:
refactor: UserForm Builder 구현

AlbertImKr committed:
refactor: User.Builder에서 ID를 필수 값으로 수정

AlbertImKr committed:
refactor: Mapping 기능을 User에서 ProfileForm으로 이동

AlbertImKr committed:
refactor: Mapping 기능을 User에서 UserForm으로 이동

AlbertImKr committed:
refactor: Mapping 기능을 User에서 ProfileSettingForm으로 이동

AlbertImKr committed:
refactor: UserController에서 Service를 이용해서 Repository의 의존성을 제거

AlbertImKr committed:
refactor: PostController에서 Service를 이용해서 Repository의 의존성을 제거

AlbertImKr committed:
refactor: UsersRepository에서 UserRepository으로 변경

AlbertImKr committed:
refactor: 네이밍 PostRepository에서 PostRepository으로 변경

AlbertImKr committed:
test: Assertions를 static으로 import로 변경

AlbertImKr committed:
test: 상수처리

AlbertImKr committed:
fix: PostController에서 필요없는 PostRepository를 삭제

AlbertImKr committed:
refactor: 에러 메시지를 messages.properties과 연결

AlbertImKr committed:
refactor: LocalValidator에 메시지소스를 등록

AlbertImKr committed:
refactor: 게시글 작성 URl를 /post에서 /postwrite로 변경

AlbertImKr committed:
refactor: SimplePostForm 매핑 기능을 SimplePostForm으로 이동

AlbertImKr committed:
refactor: PostForm에서 Post 생성 로직을 service에서 PostForm으로 이동

AlbertImKr committed:
refactor: JoinForm에서 Post 생성 로직을 service에서 JoinForm으로 이동

AlbertImKr committed:
fix: 필요없는 parameter를 삭제

AlbertImKr committed:
test: 테스트 수정

/post Url -> /postwrite로 변경

AlbertImKr committed:
test: 테스트 수정

AlbertImKr committed:
style: 필요 없는 import 제거 및 네이밍 변경

AlbertImKr committed:
chore: h2와 Spring-boot-start-jdbc 추가

AlbertImKr committed:
feat: h2 DB 추가

AlbertImKr committed:
style: URL /post/new -> /posts/new 변경

AlbertImKr committed:
chore: h2 Version를 Spring boot가 관리하도록 Version 삭제

AlbertImKr committed:
test: 테스트 수정

AlbertImKr committed:
refactor: URL /posts/new (Method.post) -> /posts로 변경

Method.Post 자체의 의미를 고려하지 않았다.

AlbertImKr committed:
refactor: URL /posts/new (Method.get) -> /posts/form로 변경

view 이름과 일치해서 form이 조금 더 적합하다.

AlbertImKr committed:
refactor: WebMmvConfigurer 의 addViewController 사용하여 /posts/new 및 /posts/form 매칭

AlbertImKr committed:
test: 테스트 수정

AlbertImKr committed:
refactor: User의 profileForm setting 메서드를 ProfileForm으로 이동

AlbertImKr committed:
test: 오류된 테스트 코드 수정

AlbertImKr committed:
style: format code 정리

AlbertImKr committed:
feat: schema.sql 및 data.sql 추가

AlbertImKr committed:
feat: JDBC SimpleJdbcInsert user save 기능 구현

AlbertImKr committed:
refactor: email를 unique로 변경

AlbertImKr committed:
refactor: UserRepostiry에서 remove 삭제하고 update로 기능을 추가

AlbertImKr committed:
feat: UPDATE_QUERY 구현

AlbertImKr committed:
style: 네이밍 변경

AlbertImKr committed:
style: QUERY문을 대분자로 변경

AlbertImKr committed:
feat: UserRepository에서 findById query 추가

AlbertImKr committed:
fix: UserRepository의 findById query 수정

AlbertImKr committed:
feat: UserRepository findByEmail query 구현

AlbertImKr committed:
feat: UserRepository containEmail query 구현

AlbertImKr committed:
fix: UserRepository save 수정

AlbertImKr committed:
fix: UserRepository getAllMembers 구현

AlbertImKr committed:
style: email 오타 수정

AlbertImKr committed:
fix: Blank도 함께 가져오는 오류를 수정

AlbertImKr committed:
fix: 기존 User 객체의 ID 사용하고 있는 부분의 오류 수정

AlbertImKr committed:
style: 상수 처리

AlbertImKr committed:
refactor: UserProfile update 개선

AlbertImKr committed:
refactor: 필용 없는 AtomicLong() 삭제

AlbertImKr committed:
refactor: User Builder ID의 final 제거

AlbertImKr committed:
style: 필용 없는 import 제거

AlbertImKr committed:
fix: 로직이 잘못된 user update를 수정

AlbertImKr committed:
refactor: 받아온 값의 spacer를 trim()으로 제거

AlbertImKr committed:
refactor: 테스트와 실제 구동의 DB를 나눔

AlbertImKr committed:
test: test 수정

AlbertImKr committed:
feat: findById 와 findByEmail 에외 처리 구현

AlbertImKr committed:
feat: enum으로 에러 관리를 구현

AlbertImKr committed:
refactor: 에러 페이지 정리

AlbertImKr committed:
feat: logger 구현

AlbertImKr committed:
style: logger에 final 추가

AlbertImKr committed:
refactor: /users를 requestMapping에 올려서 URL 간소화

AlbertImKr committed:
refactor: users의 FieldName 분리

AlbertImKr committed:
style: 필요 없는 import 와 code format을 했다

AlbertImKr committed:
refactor: id 가 존재하지 않을 때 일괄 로 에러 처리하여 필요 없는 로직을 삭제

AlbertImKr committed:
fix: email unique로 인한 검증 로직 추가

AlbertImKr committed:
fix: containEmail 결과 null인 경우를 고려하여 검증 추가

AlbertImKr committed:
refactor: 에러 페이지 수정

AlbertImKr committed:
refactor: LoginFormValidator 구현

AlbertImKr committed:
refactor: LoginFormValidator에서 UserRepostory에서 UserService를 변경

AlbertImKr committed:
refactor: JoinFormValidator 구현

AlbertImKr committed:
refactor: Repository의 프로필 업로드 시 이메일 중복 체크 기능르 Service로 이동

AlbertImKr committed:
style: code format 실행

AlbertImKr committed:
refactor: findById의 return 값을 Optional에서 User로 변경 그 대신 trow error

AlbertImKr committed:
style: 에러 출력 형식 변경

AlbertImKr committed:
style: 간격 수정

AlbertImKr committed:
feat: Repository 모든 로직에 에러 구현

AlbertImKr committed:
refactor: 필요없는 생성자 제거

AlbertImKr committed:
style: Users 패키지 정리

AlbertImKr committed:
refactor: User save할 때 키를 return 하도록 수정

AlbertImKr committed:
refactor: findById의 return 값을 다시 Optional로 변경 Service에서 throw error로 수정

AlbertImKr committed:
refactor: LoginFormValidator 필요없다고 판단하여 삭제

AlbertImKr committed:
refactor: user 생성 Post URL를 /users/join에서  /users로 변경

AlbertImKr committed:
style: 필요 없는 import 와 변수 제거

AlbertImKr committed:
refactor: 필요 없는 메서드 클래스를 제거

AlbertImKr committed:
refactor: Controller logging 구현

AlbertImKr committed:
style: 패키지 정리

AlbertImKr committed:
refactor: JoinForm Setter 제거

AlbertImKr committed:
refactor: LoginForm Setter 제거

AlbertImKr committed:
refactor: ProfileForm Setter 제거

AlbertImKr committed:
refactor: PostForm setter 제거

AlbertImKr committed:
refactor: MainController "/" 제거

AlbertImKr committed:
refactor: WebConfig 삭제 GetMapping 추가

Postform으로 값을 받지 못하여 수정

AlbertImKr committed:
refactor: SimplePostForm Setter 제거

AlbertImKr committed:
fix: URL 오류 수정

AlbertImKr committed:
refactor: form 페이지 수정

AlbertImKr committed:
refactor: 필요없는 메서드 삭제

AlbertImKr committed:
refactor: schema.sql User와의 관계를 해제

연결된 비지로직 없어서 일단 연결을 끊고 나중에 추가 할 예정입니다.

AlbertImKr committed:
refactor: Post 객체 빌드의 id를 필수 값에서 일반 값으로 변경

AlbertImKr committed:
style: 패키지 정리 및 네이밍 수정

AlbertImKr committed:
refactor: Post.Builder 변경으로 인한 수정

AlbertImKr committed:
refactor: PostRepository H2 DB로 재구성

AlbertImKr committed:
feat: Post 도메인에 필요한 에러 코드 추가

AlbertImKr committed:
test: 테스트 수정

AlbertImKr committed:
style: 패키지 재정리

AlbertImKr committed:
style: 상수 처리

AlbertImKr committed:
style: 상수 처리 및 정리

AlbertImKr committed:
style: 에러 코드 정리

AlbertImKr committed:
feat: logger 추가

AlbertImKr committed:
feat: logger 추가

AlbertImKr committed:
refactor: logger level 조절

AlbertImKr committed:
style: 코드 정리

AlbertImKr committed:
style: 패키지 네이밍 변경

AlbertImKr committed:
refactor: Post 저장후 return하는 id 타입을 int 에서 Long으로 변경

AlbertImKr committed:
refactor: method 분리

AlbertImKr committed:
refactor: method 분리

AlbertImKr committed:
fix: query 에러를 정리

동일한 rowMapper 사용해서 문제가 발생

AlbertImKr committed:
fix: save return 값을 int 로 casting으로 변경

H2 DB에서 return 값이 Interger입니다.

AlbertImKr committed:
style: createNewUser에서 saveNewUser으로 네이밍 변경

AlbertImKr committed:
refactor: PostRepository에서 logger 수정

AlbertImKr committed:
refactor: 메서드 분리 및 상수 처리

AlbertImKr committed:
refactor: inline 처리

AlbertImKr committed:
docs: URL Convention 수정

janeljs commented on README.md:
```diff
@@ -1,95 +1,35 @@
+| /users                 | post   | 회원 가입 요청                          |
+| /users/{userId}        | get    | 회원 프로필 페이지 요청                     |
+| /users/{userId}/update | get    | 회원 프로필 설정 페이지 요청                  |
+| /users/{userId}/update | put    | 회월 프로필 설정 요청                      |
```
https://github.com/codesquad-members-2023/be-java-cafe-max/pull/1/files#r1155154437
질문에 대한 답변을 받지 못하였는데요~ㅎㅎ `update`를 url에 포함시키신 이유가 궁금합니다.
폼 조회 페이지와의 url 통일성 때문인가요?

AlbertImKr replied:
🤦‍♂️🤦‍♂️Comment를 놓쳤습니다. 제 실수 입니다.
- 전체적으로 한번 더 확인 해보겠습니다.

AlbertImKr replied:
💡이렇게 개선하겠습니다.

| NO. | Pagename  | HttpRequset Method  | 수정전 URL | 수정후 URL | 이유 |
|--|--|--|--|--|--|
|1|  개인 프로필 페이지 |  GET | /users/{userId} | /users/{userId}/profile | /users/{userId}는 개인 정보 페이지로 구성하려고 수정했습니다. |
|2| 개인 프로필 수정 페이지 | Get  | /users/{userId}/update | /users/{userId}/profile/edit |
|3| 개인 프로필 수정 요청    | Put  | /users/{userId}/update | /users/{userId}/profile      |

janeljs commented on README.md:
```diff
@@ -1,95 +1,35 @@
+| /                      | get    | 모든 게스글 표시되는 메인 페이지 요청             |                    
+| /posts/{postId}        | get    | 게시글의 페이지 요청                       |
+
+## [Commit Log Guidelines](https://github.com/naver/egjs/wiki/Commit-Log-Guidelines)
+- feat: A new feature
+- fix: A bug fix
+- docs: Documentation only changes
+- style: Changes that do not affect the meaning of the code. Such as white-space, formatting, missing semi-colons, etc... It also possible to change JSHint, JSCS rules of the code.
+- refactor: A code change that neither fixes a bug nor adds a feature
+- test: Adding missing tests. Changing tests.
+- demo: Adding missing demos. Changing demos.
+- chore: Changes to the build process or tools and libraries such as documentation generation
```
👍🏻

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserService.java:
```diff
@@ -1,52 +1,65 @@
+        return userRepository.findByEmail(email);
+    }
+
+    public boolean containEmail(String email) {
```
리뷰 반영이 안된 것 같아요~ㅎㅎ https://github.com/codesquad-members-2023/be-java-cafe-max/pull/1/files#r1155157429

AlbertImKr replied:
수정하겠습니다. 🕵️

janeljs commented on src/main/java/kr/codesqaud/cafe/account/JoinFormValidator.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public void validate(Object target, Errors errors) {
+        JoinForm joinForm = (JoinForm) target;
+        if (userService.containEmail(joinForm.getEmail())) {
+            errors.rejectValue(EMAIL, "error.email.duplicate");
```
JoinFormValidator로 validation 로직을 분리해주셨군요~ㅎㅎ 지금은 이메일 검증만 있어서 조금 과한 느낌도 있는 거 같은데 추후에 검증 로직이 많이 생기면 관리가 쉬워지겠네요 👍🏻

AlbertImKr replied:
- Validator을 한번 사용해봤습니다.  
- Validator은 역할이 어디 까지인지 고민하고 있습니다.
     - 예를 들면 로그인 로직에서
          - 로그인시 입력 타입을 검증한다.
          - 존재하는 아이디 인지 검증한다.
          - 비밀번호가 체크한다. 
    - 이 모든 것을 Validator에서 모두 검증하여도 되나요?🧐

janeljs replied:
음..이 부분은 정답은 없습니다. 말씀하신 로직들 다 validator에서 처리해도 괜찮고, 일부는 서비스에서 일부는 validator에서 하셔도 괜찮습니다.
복잡한 비즈니스 로직이 많아서 서비스가 엄청 비대해져 로직을 관리하기 어려울 때가 있는데요.
그럴 때 CQRS로 서비스를 분리하여도 괜찮고, validate 관련 로직은 validator로 넣어도 괜찮습니다.

참고로 validator 인터페이스가 도입될 때 어플리케이션의 모든 layer에서 사용될 수 있게 만들어졌다고 하네요ㅎㅎ
> Spring has come up with a Validator interface that is both basic and eminently usable in every layer of an application.

https://docs.spring.io/spring-framework/docs/2.5.5/reference/validation.html

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+import java.util.Optional;
 
 @Controller
+@RequestMapping("/users")
```
nit: 공통부분을 `@RequestMapping`으로 분리해주셨군요~ 이렇게 하면 중복 코드가 줄어드는 장점이 있기는 한데요.
인텔리제이에서 api 전체 url로 검색할 때 나오지 않고, path 가독성이 오히려 떨어지는 것 같아 개인적으로는 선호하지 않습니다ㅎㅎ

AlbertImKr replied:
저도 분리하는 것이 조금 가독성이 떨어진다고 생각은 들었습니다. 다른 리뷰어 분의 코멘트를 보고 수정했습니다. 저 다시 돌려 놓겠습니다.🤩🤩🤩

jinny-l replied:
> nit: 공통부분을 `@RequestMapping`으로 분리해주셨군요~ 이렇게 하면 중복 코드가 줄어드는 장점이 있기는 한데요. 인텔리제이에서 api 전체 url로 검색할 때 나오지 않고, path 가독성이 오히려 떨어지는 것 같아 개인적으로는 선호하지 않습니다ㅎㅎ

@janeljs 
api 전체 url로 검색할 때가 shift 두번 눌렀을 때를 말씀하시는 것일까요?

janeljs replied:
앗 이부분은 cmd + shift + f 로 검색했을 때를 말씀드린 건데요ㅎㅎ shift 두 번 눌러서 전체검색 할때는 잘 나옵니다ㅎㅎ

&rarr; 저는 개인적으로 너무 길지 않다면 full path를 쓰는 것이 가독성에 더 좋다고 생각하지만 온전히 취향의 영역이라 @RequestMapping으로 중복코드를 제거해주시는 것도 좋은 방법인 것 같습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+import java.util.List;
+import java.util.Optional;
 
 @Controller
```
`@RestController`와 `@Controller`의 차이점에 대하여 학습해보세요~

AlbertImKr replied:
- @RestController는 @Controller + @ResponseBody 이라고 알고 있습니다. 
- String 반환하는 Controller입니다. 
- 조금 더 찾아보고 정리하고 알아보겠습니다. 📚

janeljs replied:
네~ㅎㅎ spring mvc에서 DispatcherServlet으로 요청이 들어와서 사용자에게 반환되기까지 어떤 과정이 있는지도 같이 학습해보세요ㅎㅎ

> String 반환하는 Controller입니다.

이거는... `@Controller`에서 view name 반환한다고 말씀주신 것 맞죠..?ㅎㅎㅎ

RestController에 왜 "Rest"가 붙었는지와 응답 포맷도 함께 생각해보세요~

참고: https://www.geeksforgeeks.org/difference-between-controller-and-restcontroller-annotation-in-spring/

AlbertImKr replied:
> 이거는... @Controller에서 view name 반환한다고 말씀주신 것 맞죠..?ㅎㅎㅎ

문자열을  반환한다고 표현하고 싶었습니다.  🫣
- JSON형식
- 그냥 String 문자열

janeljs replied:
@Controller에서 string 문자열 말고 ModelAndView를 리턴할 때도 있어서요ㅎㅎ 결국에는 string도 ModelAndView로 담기게 spring이 처리해주지만요ㅎㅎㅎ

참고: https://stackoverflow.com/questions/25988442/what-is-the-difference-between-return-modelandview-and-return-string-in-spring-m

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+                        error.getDefaultMessage()));
+    }
+
+    @InitBinder("joinForm")
```
`@InitBinder`를 사용하셨군요 👍🏻 joinForm 지정도 잘해주셨네요ㅎㅎ

보너스로ㅎㅎ 아래와 같이 선언해주면 어떤 일이 일어날까요?(힌트 키워드: setter)
```java
@ControllerAdvice
public class WebControllerAdvice {

    @InitBinder
    public void initBinder(WebDataBinder binder) {
        binder.initDirectFieldAccess();
    }
}
```

AlbertImKr replied:
- DataBinder 의 기본 설정은 `initBeanPropertyAccess()` 입니다. JavaBean property access로 객체를 초기화한다. 즉 setter 사용하여 초기화 한다.
- `initDriectFiedlAccess()` 로 설정하면 ~~reflection을 통해서~~ 직접 필드를 access하여 초기화 한다. 즉 setter를 사용하지 않고 초기화 한다.
- 장점:
      1.  노출하고 싶은 부분만 외부에 노출
      2. 상태는 잘 캡슐화되어 있습니다.
      3. ~~인터페이스, 추상 클래스 또는 매핑된 슈퍼클래스에서 getter 및 setter를 선언하고 구체적인 서브클래스에서 override할 수 있습니다. 임시 계약을 정의하는 것이 더 쉽습니다.~~
- 단점:
     -  컨테이너는 값 주입을 위해 게터와 세터를 사용하지 않습니다. injection-process는 direct field injection으로 디버깅할 수 없습니다.

janeljs replied:
> DataBinder 의 기본 설정은 initBeanPropertyAccess() 입니다. JavaBean property access로 객체를 초기화한다. 즉 setter 사용하여 초기화 한다.
> initDriectFiedlAccesss() 로 설정하면 reflection을 통해서 직접 필드를 access하여 초기화 한다. 즉 setter를 사용하지 않고 초기화 한다.

굳굳 잘 학습하셨네요~ㅎㅎ

janeljs replied:
>3. 인터페이스, 추상 클래스 또는 매핑된 슈퍼클래스에서 getter 및 setter를 선언하고 구체적인 서브클래스에서 override할 수 있습니다. 임시 계약을 정의하는 것이 더 쉽습니다.

이 부분은 제가 잘 이해를 못했는데 임시계약을 정의하는게 더 쉽다는게 무슨 뜻인가요?ㅎㅎ
또한 인터페이스, 추상 클래스 또는 수퍼클래스의 메서드를 구현체 또는 서브 클래스에서 override할 수 있는 건 해당 설정을 하지 않아도 적용되는 부분 같은데 왜 장점에 넣어주셨나 궁금합니다.

AlbertImKr replied:
initDirectFieldAccess() 만 가능하다고 생각하였는데 찾아보니까 그런 내용이 아닌것 같습니다. 🤦‍♂️

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+    }
+
+    @PostMapping
+    public String addUser(@Valid JoinForm joinForm, BindingResult bindingResult) {
```
`@Validated`와는 어떤 차이점이 있을까요~?

AlbertImKr replied:
@Valid 는 javax의 에노테이션으로서 java표준의 에노테이션입니다. @Validated 는 Spring의 에노테이션입니다. 지금 상상할수 없겠지만 Spring framwork를 자른 framwork로 변경할 때 @Valid를 사용 했으면 이점이 있습니다.

janeljs replied:
`@Validated`에서만 지원되는 기능들(e.g. groups)도 있기는 한데요ㅎㅎ 잘 사용하지는 않아서 지금 정도의 학습도 좋은 것 같습니다. 👍🏻

AlbertImKr replied:
`@Validated` 지원되는 기능들도 한번 찾아보겠습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserService.java:
```diff
@@ -1,52 +1,65 @@
+        this.userRepository = userRepository;
+    }
+
+    public int saveNewUser(JoinForm joinForm) {
```
new는 없어도 될 것 같아요~ https://github.com/codesquad-members-2023/be-java-cafe-max/pull/1/files#r1155154619 코멘트 보고 save로 바꾸신 것 같은데요ㅎㅎ `join` 등을 생각하고 드린 코멘트이기는 했습니다.
그런데 이거는 Service와 Repository 간 네이밍을 통일하게 팀 컨벤션을 정할 수도 있는 거고해서 통일성만 갖춰주신다면 `saveUser`도 괜찮을 것 같아요ㅎㅎ

AlbertImKr replied:
🤦‍♂️🤦‍♂️🤦‍♂️ 통일하도록 하겠습니다.  GitHub가 리뷰시스템을 아직 익숙하지 않아서 많은 Comment를 확인 못했습니다.
- saveNewUser -> saveUser로 변경했습니다. (네이밍이 제일 어렵습니다.)

janeljs commented on src/main/java/kr/codesqaud/cafe/account/exception/SaveUserFailedException.java:
```diff
@@ -0,0 +1,11 @@
+import kr.codesqaud.cafe.exception.CustomException;
+import kr.codesqaud.cafe.exception.ErrorCode;
+
+public class SaveUserFailedException extends CustomException {
```
예외의 이름들이 조금 어색한 것 같아요. 여러 사례를 찾아보고 개선해보세요~

AlbertImKr replied:
고민하고 개선해보겠습니다.

janeljs commented on src/main/resources/messages_ko.properties:
```diff
@@ -0,0 +1,8 @@
+error.email.notExist=\uC774\uBA54\uC77C\uC774 \uC874\uC7AC\uD558\uC9C0 \uC54A\uC2B5\uB2C8\uB2E4.
```
흠.. 인코딩 설정 어떤 거로 되어있으세요?

AlbertImKr replied:
- 국제화 resource를 사용했는데 한글이 깨져서 Intellij 에서 encoding UTF-8로 설정했고 Transparent native-to-ascii conversion했습니다. 코드가 이렇게 올라가는 것을 이제 알았습니다.🧐🧐
<img width="579" alt="스크린샷 2023-04-07 오후 5 06 00" src="https://user-images.githubusercontent.com/99056666/230569246-f93b6d5d-4b77-49fc-87d3-dd3f01b4ce8e.png">

janeljs replied:
`transparent native-to-ascii conversion` 옵션을 적용하지 않아도 처음 파일 입력 시 인코딩이 utf-8로 지정되어 있으셨다면 push본도 한글로 잘 보이고 로컬에서도 한글로 잘 보일 것 같은데요ㅎㅎ 해당 옵션을 적용하신 이유가 있었을까요?!

AlbertImKr replied:
- 처음 국제화 메시지를 사용 할 때 한글이 깨져서 방법을 찾아봤습니다. 
- Transparent native-to-sacii conversion의 의미를 제대로 이해하지 못했습니다. 👍

janeljs commented on src/main/java/kr/codesqaud/cafe/account/dto/ProfileSettingForm.java:
```diff
@@ -0,0 +1,87 @@
+    @Pattern(regexp = "^(.*[a-z]+.*[1-9]+가.*)|(.*[1-9]+.*[a-z]+.*)$", message = "{error.password.pattern}")
+    private final String password;
+
+    public ProfileSettingForm(String nickname, String email, String password) {
+        this.nickname = nickname;
+        this.email = email;
+        this.password = password;
+    }
```
이 생성자는 어디 쓰이고 있나요?

AlbertImKr replied:
- form에서 Setter를 모두 제거했고 Builder로 생성자를 구현하여 Form에서 받아온 정보를 객체로 DataBinder할 때 사용 하고 있습니다.

janeljs replied:
빌더에서는 아래 생성자만 사용되는 것 같아 여쭤보았습니다~ㅎㅎ
```java
  private ProfileSettingForm(Builder builder) {
        this.email = builder.email;
        this.nickname = builder.nickname;
        this.password = builder.password;
    }
```

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+            loggingError(bindingResult);
+            return "account/profileUpdate";
+        }
+        User user = userService.findById(userId);
+        if (userService.isDuplicateEmail(user.getEmail(), profileSettingForm.getEmail())) {
+            loggingError(bindingResult);
+            bindingResult.rejectValue(EMAIL, "error.email.duplicate");
+            return "account/profileUpdate";
+        }
+        if (!userService.checkPassword(profileSettingForm.getPassword(), user.getPassword())) {
+            loggingError(bindingResult);
+            bindingResult.rejectValue(PASSWORD, "error.password.notMatch");
+            return "account/profileUpdate";
+        }
+        userService.update(profileSettingForm, userId);
```
가독성도 안좋고 서비스에서 처리해야 할 로직이 Controller에 있는 것 같아요.
그리고 불필요한 DB 조회도 발생하구요.
`userService.isDuplicateEmail(form.getEmail())`, `userService.checkPassword(form.getPassword())` 이런식으로 되도록 리팩토링해보는 건 어떨까요? 

+ 비밀번호 암호화에 대해서도 나중에 생각해보세요ㅎㅎ

AlbertImKr replied:
- `userService.isDuplicateEmail(form.getEmail())` 
      -  이 부분은 DB에서 email조회 하면 될 것 같습니다.
 - `userService.checkPassword(form.getPassword())`
      -  이 부분은 어떻게 해결 할지 모르겠어요.  힌트 주실수 있을 까요 ?😭

AlbertImKr replied:
- 이렇게 수정했습니다.
- DB 조회는 두번은 해야 한다고 생각합니다.
     - Email 중복 체크
     - Password 일치 체크

Controller
``` java
        if (bindingResult.hasErrors()) {
            loggingError(bindingResult);
            return "account/profileEditForm";
        }
        userService.update(profileEditForm, userId, bindingResult);
        if (bindingResult.hasErrors()) {
            return "account/profileEditForm";
        }
        return "redirect:/users/{userId}/profile";
```
Service
``` java
    public void update(ProfileEditForm profileEditForm, Long userId, BindingResult bindingResult) {
        User user = userRepository.findById(userId);
        if (!user.isSameEmail(profileEditForm.getEmail()) && isDuplicateEmail(profileEditForm.getEmail())) {
            bindingResult.rejectValue(EMAIL, "error.email.duplicate");
            loggingError(bindingResult);
            return;
        }
        if (!user.isSamePassword(profileEditForm.getPassword())) {
            bindingResult.rejectValue(PASSWORD, "error.password.notMatch");
            loggingError(bindingResult);
            return;
        }
        userRepository.update(profileEditForm.setUser(findById(userId)));
    }
```

janeljs replied:
`불필요한 DB 조회`는 아래 부분에 대한 코멘트였습니다ㅎㅎ
```java
  User user = userService.findById(userId);
        if (userService.isDuplicateEmail(user.getEmail(), profileSettingForm.getEmail()))
```
```java
public boolean isDuplicateEmail(String targetEmail) {
        return userRepository.findByEmail(targetEmail).isPresent();
    }
```

아래와 같이 개선하신 것 확인했는데요ㅎㅎ
```java
        if (userService.isDuplicateEmail(profileSettingForm.getEmail())) {
            bindingResult.rejectValue(EMAIL, "error.email.duplicate");
            loggingError(bindingResult);
            return "account/profileUpdate";
        }
        if (!userService.isSamePassword(userId,profileSettingForm.getPassword())) {
            bindingResult.rejectValue(PASSWORD, "error.password.notMatch");
            loggingError(bindingResult);
            return "account/profileUpdate";
        }
```
service에서 중복검사를 처리하고, 예외를 던진 후 해당 예외를 잡아 bindingResult에서 처리하거나, exception handler에서 처리하도록 개선해주세요ㅎㅎ

janeljs replied:
앗 이부분 제가 깊게 생각하지 않고 제안을 드린 것 같습니다 😢 
위에 해주신 것처럼 user에게 물어보는 형태로 `user.isSamePassword`와 같이 개선해주신 것 좋습니다 👍🏻 

+ 이 부분도 service에서 중복검사를 처리하고, 예외를 던진 후 해당 예외를 잡아 controller에서 bindingResult로 처리하거나, exception handler에서 처리하도록 개선해주세요ㅎㅎ

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+    }
+
+    @GetMapping
+    public String showUsers(Model model) {
+        List<UserForm> allUserForm = userService.getAllUsersForm();
+        model.addAttribute(USERS, allUserForm);
+        return "account/members";
+    }
```
여러군데 걸쳐서 있는 것 같은데 user와 member를 혼용해서 쓰시는 이유가 있나요?

AlbertImKr replied:
통일 하도록 하겠습니다.

janeljs commented on src/test/java/kr/codesqaud/cafe/account/UserControllerTest.java:
```diff
@@ -15,17 +14,35 @@
 
 @AutoConfigureMockMvc
 @SpringBootTest
+@Transactional
```
테스트에 `@Transactional`을 붙이면 어떻게 동작하나요~?

AlbertImKr replied:
- 테스트 후 Rollback 을 합니다. 실행된 query가 모두 취소된다고 알고 있습니다.

janeljs replied:
롤백을 해주기는 하는데요ㅎㅎ auto increment 컬럼은 이미 증가한 값 그대로 유지될 거예요ㅎㅎ 한 번 확인해보시고 알려주세요~ㅎㅎ

AlbertImKr replied:
- 트랜잭션을 롤백해도 여전히 기본 키가 증가합니다.
- auto increment  자동 증가 값을 생성하는 데 사용되는 잠금 모드와 관련이 있습니다. 
- Insert문에 atuo increament의 값의 중복을 방지하기 위하여 할당하여 실행한다. insert Transation의 속하지 않아 rollback할 수 없다고 이해했습니다.

> https://stackoverflow.com/questions/449346/mysql-auto-increment-does-not-rollback

janeljs commented on src/test/java/kr/codesqaud/cafe/account/UserControllerTest.java:
```diff
@@ -184,42 +197,32 @@ void setUserProfileFailedByPassword() throws Exception {
-		User user = saveAndGetUserJack();
-		String mail = "jerry@email.com";
-		String jerry = "jerry";
+		saveAndGetUserJack();
 		mockMvc.perform(put("/users/20/update")
```
여기는 왜 userId 값이 하드코딩 되어있나요?

AlbertImKr replied:
- 저장한 userId 받고 ++userId 를 사용하여 20을 대체했습니다.

janeljs replied:
뭔가 자연스러운 방법은 아닌 것 같습니다.
차라리 static 상수로 `NON_EXISTING_USER_ID = 2` 이런식으로 선언해주시고 사용하시는 건 어떨까요?

AlbertImKr replied:
좋은 방법인 것 같습니다. 👍

janeljs commented on src/main/java/kr/codesqaud/cafe/post/PostRepository.java:
```diff
@@ -0,0 +1,108 @@
+                    getPostRowMapper(),
+                    postId));
+        } catch (DataAccessException e) {
+            logger.info("[ Message = {} ],[ Id = {} ]", NO_SUCH_POST_ID_CODE.getMessage(), postId);
```
로깅 level info가 적합할까요?

AlbertImKr replied:
- Id 없을 때 발생하는 에러이라고 생각했습니다. 그래서 Id 가 없을 수 있는 것은 매우 정상적이 일이라고 생각해서 info를 사용했습니다. 
- 각 레벨의 개념은 찾아서 봤는데 현 단계 info 이상의 레벨을 사용해야 하는 필요성을 느끼지 못했습니다.
- 항상 기준을 잡는 것이 제일 힘든 부분이라고 생각합니다.😭

janeljs replied:
본인만의 타당한 이유를 가지고 설정해 주신 것 좋습니다 💯 
말씀주신대로 개발자가 이미 예상할 수 있는 예외 상황이라서 info 레벨 정도면 충분한 것 같네요ㅎㅎ

+ 로깅 레벨에 대한 다양한 토론이 있어 첨부드립니다.
참고: https://stackoverflow.com/questions/2031163/when-to-use-the-different-log-levels

jinny-l commented:
앨버트 AWS 배포하는 거 별로 안 어려우셨나요?? 나중에 하신다더니 뚝딱하셨네요.

저는 새로운 개념에 대해 학습할 때 많이 헤매고 시간도 오래 걸리는데, 
앨버트는 항상 뚝딱뚝딱 빨리 해내시는 것 보면 대단하다고 생각이 들어요.
특히 이번주가 저는 많이 헤매서 많이 못한 것 같은데 앨버트는 정말 많이 구현하셔서 대단합니다...

janeljs commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,6 @@
+spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.url=jdbc:h2:mem:test
+spring.datasource.username=sa
+spring.datasource.password=
```
jasypt를 이용해서 중요 정보를 암호화해보세요~

AlbertImKr replied:
넵. 암호화 해보겠습니다.🧑‍💻

janeljs commented on src/main/java/kr/codesqaud/cafe/exception/ErrorCode.java:
```diff
@@ -0,0 +1,47 @@
+
+import org.springframework.http.HttpStatus;
+
+public enum ErrorCode {
```
ㅎㅎ 고생하셨어요~ 💯

AlbertImKr replied:
감사합니다. 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️

janeljs commented on src/main/java/kr/codesqaud/cafe/exception/ErrorCode.java:
```diff
@@ -0,0 +1,47 @@
+
+public enum ErrorCode {
+
+    INVALID_USER_ID_CODE(HttpStatus.BAD_REQUEST, "US001", "접근 불가능한 User 아이디입니다.", "error/custom"),
+    NO_SUCH_USER_ID_CODE(HttpStatus.NOT_FOUND, "UD001", "해당 User 아이디가 존재하지 않습니다.", "error/custom"),
+    NO_SUCH_EMAIL_CODE(HttpStatus.NOT_FOUND, "UD002", "해당 User 이메일이 존재하지 않습니다.", "error/custom"),
+    SAVE_USER_FAILED_CODE(HttpStatus.BAD_REQUEST, "UD003", "해당 User 정보를 저장하는데 실패 했습니다.", "error/custom"),
+    GET_ALL_USERS_FAILED_CODE(HttpStatus.BAD_REQUEST, "UD004", "User 전체 조회 실패 했습니다.", "error/custom"),
+    UPDATE_USER_FAILED_CODE(HttpStatus.BAD_REQUEST, "UD005", "User 업데이트 실패 했습니다.", "error/custom"),
+    INVALID_POST_ID_CODE(HttpStatus.BAD_REQUEST, "PS001", "접근 불가능한 Post 아이디입니다.", "error/custom"),
+    SAVE_POST_FAILED_CODE(HttpStatus.BAD_REQUEST, "PD001", "해당 Post 정보를 저장하는데 실패 했습니다.", "error/custom"),
+    NO_SUCH_POST_ID_CODE(HttpStatus.NOT_FOUND, "PD002", "해당 Post 아이디가 존재하지 않습니다.", "error/custom"),
+    GET_ALL_POSTS_FAILED_CODE(HttpStatus.NOT_FOUND, "PD003", "Post 전체 조회 실패 했습니다.", "error/custom"),
+    NO_SUCH_POST_TITLE_CODE(HttpStatus.NOT_FOUND, "PD004", "해당 Post 타이틀이 존재하지 않습니다.", "error/custom"),
```
전부 커스텀 에러 페이지로 보내도 괜찮을까요? (서비스 화면으로 이동되어야 하는 경우는 없나요?)

AlbertImKr replied:
- Exception을 던져서 Error view 가 아닌 일반 컨트롤러 view(예: 로그인 페이지)로 연결하는 방법은 좋은 방법 일 까요?
```java
@ExceptionHandler(LoginException.class)
public ModelAndView handler(....){
          
         modelAndView.addObject("email","email 오류")
         return "users/login";
}
```

janeljs replied:
> Exception을 던져서 Error view 가 아닌 일반 컨트롤러 view(예: 로그인 페이지)로 연결하는 방법은 좋은 방법 일 까요?

해당 예외들이 발생했을 때 사용자에게 전부 커스텀 에러 페이지가 보이는게 맞는가(?)하는 의문에서 위와 같은 코멘트를 드려봤습니다.
예를 들어 로그인을 하지 않은 사용자가 특정 포스트를 수정하려고 해서 수정 불가능한 post라는 예외가 발생했다고 가정했을 때는 에러페이지를 보여주기보단 로그인 페이지로 보내주는게 좋지 않나 라는 생각을 해보았습니다ㅎㅎ

위와 같은 상황에 대한 처리는 따로 되어있고, 커스텀 에러페이지를 보여주는 것이 더 나은 상황이라면 상황에 맞는 에러 페이지를 적절하게 구현하여 지금과 같이 enum에서 관리하고 있는 viewname을 리턴하도록 처리해주는 것도 좋은 방법입니다 👍🏻

janeljs commented on src/main/java/kr/codesqaud/cafe/post/PostRepository.java:
```diff
@@ -0,0 +1,108 @@
+    public static final String COLUMN_TEXT_CONTENT = "TEXT_CONTENT";
+    public static final String COLUMN_CREATE_DATETIME = "CREATE_DATETIME";
+    public static final String QUERY_SELECT_ALL = "SELECT * FROM POST";
+    public static final String QUERY_FIND_BY_ID = "SELECT POST_ID , NICKNAME, TITLE, TEXT_CONTENT, CREATE_DATETIME FROM POST WHERE POST_ID = ?";
+    public static final String QUERY_FIND_BY_TITLE = "SELECT POST_ID , NICKNAME, TITLE, TEXT_CONTENT, CREATE_DATETIME FROM POST WHERE TITLE = ?";
```
named parameter에 대해 공부해보시는 것도 좋을 것 같네요ㅎㅎ

AlbertImKr replied:
- named parameter도 적용했습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/config/ValidatorConfig.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.validation.beanvalidation.LocalValidatorFactoryBean;
+
+public class ValidatorConfig {
+    MessageSource messageSource;
```
접근제어자가 누락된 것 같네요~

AlbertImKr replied:
🤦‍♂️🤦‍♂️🤦‍♂️ 수정하겠습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/post/PostRepository.java:
```diff
@@ -0,0 +1,108 @@
+    public static final String COLUMN_NICKNAME = "NICKNAME";
+    public static final String COLUMN_TEXT_CONTENT = "TEXT_CONTENT";
+    public static final String COLUMN_CREATE_DATETIME = "CREATE_DATETIME";
+    public static final String QUERY_SELECT_ALL = "SELECT * FROM POST";
```
쿼리에서 `*`를 사용하는 것은 지양하는게 좋습니다. 컬럼을 명시해주세요.

AlbertImKr replied:
- 수정하겠습니다.🧑‍💻🧑‍💻🧑‍💻

janeljs commented on src/main/java/kr/codesqaud/cafe/post/PostRepository.java:
```diff
@@ -0,0 +1,108 @@
+    private final DataSource dataSource;
+
+    public PostRepository(DataSource dataSource) {
+        this.jdbcTemplate = new JdbcTemplate(dataSource);
```
생성자로 생성하지 않고 JdbcTemplate을 주입받을 수는 없을까요?

AlbertImKr replied:
가능합니다.
- Spring jdbc를 사용시 `@Autowired` 를 사용하면 바로 사용 가능합니다.
- ```java
    @Autowired
    private JdbcTemplate jdbcTemplate;
    @Autowired
    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;

- dataSource를 주입하지 않고 바로 생성자 주입도 가능합니다.

- ```java
    private final JdbcTemplate jdbcTemplate;
  
    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
    
    
    public UserRepository(JdbcTemplate jdbcTemplate, NamedParameterJdbcTemplate namedParameterJdbcTemplate) {
              this.jdbcTemplate = jdbcTemplate;
              this.namedParameterJdbcTemplate = namedParameterJdbcTemplate;
    }

janeljs commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,20 @@
+DROP TABLE IF EXISTS post;
+DROP TABLE IF EXISTS users;
+CREATE TABLE users
```
`user`가 좋을 것 같아요ㅎㅎ 왜그런지 이유도 생각해보세요~

AlbertImKr replied:
감사합니다. 확인 했습니다. 
- https://stackoverflow.com/questions/338156/table-naming-dilemma-singular-vs-plural-names
- 처음에 user를 사용하려고 했는데  user가 H2가 System Table이여서 사용할 수 없었습니다.   테이블 이름을 account로 변경하여 사용하겠습니다.

janeljs replied:
아하 그런 배경이 있었군요ㅠㅠ 혹시 double quote로 감싸도 안되나요?
> There is a list of keywords that can't be used as identifiers (table names, column names and so on), unless they are quoted (surrounded with double quotes). 

http://www.h2database.com/html/advanced.html

아래와 같이도 사용할 수 있다는데 키워드에서 인위적으로 제외하는게 뭔가 좋아보이지는 않네요..
```yml
spring:
  datasource:
    url: jdbc:h2:mem:testdb;NON_KEYWORDS=USER
```

AlbertImKr replied:
- 아직 배울 것이 많네요.

janeljs commented on src/main/java/kr/codesqaud/cafe/post/PostRepository.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    public int save(Post post) {
+        try {
+            SimpleJdbcInsert simpleJdbcInsert = new SimpleJdbcInsert(dataSource).withTableName(TABLE_NAME)
+                    .usingGeneratedKeyColumns(COLUMN_ID);
+            Map<String, Object> parameters = getParameters(post);
+            return (Integer) simpleJdbcInsert.executeAndReturnKey(parameters);
+        } catch (DataAccessException e) {
+            logger.debug("[ Message = {} ]", SAVE_POST_FAILED_CODE.getMessage());
+            throw new SavePostFailedException(SAVE_POST_FAILED_CODE);
+        }
```
DataAccessException을 공통으로 잡고있는 수많은 try catch와 로깅에 대하여 중복 제거를 해줄 수 있는 방법을 생각해보세요~ㅎㅎ 이미 비슷한 걸 구현하신게 있습니다.

AlbertImKr replied:
- Exception에 대하여 더 공부하고 수정해보겠습니다.😭😭
   - 혹시 추천할 Blog나 참고자료가 있을 까요?

janeljs replied:
간단하게 @ControllerAdvice를 이용한 global exception handler에서 `DataAccessException`을 잡아서 해결할 수 있을 것 같아요ㅎㅎ
```java
  // Specify name of a specific view that will be used to display the error:
  @ExceptionHandler({SQLException.class,DataAccessException.class})
  public String databaseError() {
    // Nothing to do.  Returns the logical view name of an error page, passed
    // to the view-resolver(s) in usual way.
    // Note that the exception is NOT available to this view (it is not added
    // to the model) but see "Extending ExceptionHandlerExceptionResolver"
    // below.
    return "databaseError";
  }
```
참고: https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

janeljs replied:
그리고 공부하시는 겸에ㅎㅎㅎ `SQLErrorCodeSQLExceptionTranslator`가 jdbcTemplate에서 어떤 역할을 해주었는지도 공부해보세요ㅎㅎ

https://docs.spring.io/spring-framework/docs/3.0.x/spring-framework-reference/html/jdbc.html#jdbc-SQLExceptionTranslator

AlbertImKr replied:
감사합니다. 🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️🙇🏻‍♂️

janeljs commented on src/main/java/kr/codesqaud/cafe/account/User.java:
```diff
@@ -1,67 +1,64 @@
+    private final String email;
+    private final String password;
+
+    private User(Builder builder) {
```
private으로 선언해주신 것 좋아요~

AlbertImKr replied:
감사합니다. 🙇🏻‍♂️

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserRepository.java:
```diff
@@ -0,0 +1,119 @@
+    private static final String QUERY_UPDATE = "UPDATE USERS SET NICKNAME = ?, EMAIL = ? WHERE USER_ID = ?";
+    private static final String QUERY_FIND_BY_ID = "SELECT USER_ID,EMAIL,NICKNAME,PASSWORD FROM USERS WHERE USER_ID = ?";
+    private static final String QUERY_FIND_BY_EMAIL = "SELECT USER_ID,NICKNAME,PASSWORD,EMAIL FROM USERS WHERE EMAIL = ?";
+    private static final String QUERY_CONTAINS_EMAIL = "SELECT COUNT(*) FROM USERS WHERE EMAIL = ?";
```
이렇게 되면 email이 있는지 확인하기 위해 전체를 확인해야 하는데요. 개선할 수는 없을까요?

AlbertImKr replied:
이렇게 개선했습니다. 🕵️🕵️
- ```SELECT COUNT(email) FROM USERS WHERE EMAIL = ?```

janeljs replied:
흐음.. 저는 column 말고 row의 관점에서 전체를 말씀드린 거긴 했습니다ㅎㅎ 지금 email이 있는지 없는지만 알면 되는데 몇 개 있는지까지 조회하는 쿼리가 필요할까요?

janeljs replied:
아래 글도 참고해보세요~ㅎㅎ
https://jojoldu.tistory.com/516

AlbertImKr replied:
넵 감사합니다. 😭

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserRepository.java:
```diff
@@ -0,0 +1,119 @@
+        return count != null && count != 0;
+    }
+
+    public void update(User user) {
```
이렇게 update를 따로 해줘도 되지만 upsert라는 개념도 있습니다. 참고삼아 공부해보세요ㅎㅎ

AlbertImKr replied:
- 감사합니다. 🙇🏻‍♂️🙇🏻‍♂️찾아보고 공부해보겠습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/account/UserController.java:
```diff
@@ -1,128 +1,137 @@
+            loggingError(bindingResult);
+            return "account/login";
+        }
+        Optional<User> userOptional = userService.findByEmail(loginForm.getEmail());
+        if (userOptional.isEmpty()) {
+            loggingError(bindingResult);
+            bindingResult.rejectValue(EMAIL, "error.email.notExist");
+            return "account/login";
+        }
+        User user = userOptional.get();
+        if (!user.getPassword().equals(loginForm.getPassword())) {
+            loggingError(bindingResult);
+            bindingResult.rejectValue(PASSWORD, "error.password.notMatch");
+            return "account/login";
+        }
```
컨트롤러에서 수행되어야 할 로직일까요?

AlbertImKr replied:
- ~~컨트롤러로 이동하겠습니다.~~
- 서비스로 이동하겠습니다.

AlbertImKr replied:
```java
//controller
    @PostMapping("/users/login")
    public String login(@Valid LoginForm loginForm, BindingResult bindingResult) {
        if (bindingResult.hasErrors()) {
            loggingError(bindingResult);
            return "account/login";
        }
        Long userId = userService.login(loginForm, bindingResult);
        if (bindingResult.hasErrors()) {
            return "account/login";
        }
        return "redirect:/users/" + userId + "/profile";
    }
```
```java
//service
        public Long login(LoginForm loginForm, BindingResult bindingResult) {
        Optional<User> userOptional = findByEmail(loginForm.getEmail());
        if (userOptional.isEmpty()) {
            loggingError(bindingResult);
            bindingResult.rejectValue(EMAIL, "error.email.notExist");
            return -1L;
        }
        User user = userOptional.get();
        if (!isSamePassword(user, loginForm.getPassword())) {
            loggingError(bindingResult);
            bindingResult.rejectValue(PASSWORD, "error.password.notMatch");
            return -1L;
        }
        return user.getId();
    }
```

janeljs replied:
> 컨트롤러로 이동하겠습니다.

서비스로 이동 말씀해주신 것 맞죠?!ㅎㅎ

janeljs replied:
몇 가지 코멘트 드리고 싶은게 있는데요ㅎㅎ

1. optional은 굳이 변수명에 적어주지 않으셔도 됩니다.
2. bindingResult는 form validation에만 사용해주시고, db 조회가 필요한 로직은 service에서 적절한 예외를 던져주세요.
bindingResult로 처리하면 좋은 로직들은 아래와 같습니다.
![image](https://user-images.githubusercontent.com/68000537/230940362-f3393a79-6d23-4a2d-b760-ea6891c613fe.png)
출처: https://web.archive.org/web/20161120115428/http://codetutr.com/2013/05/28/spring-mvc-form-validation/

일단 위에 첨부해주신 코드처럼 서비스에서 bindingResult가 등장하면 뭔가 이상한데..라는 생각을 해주시고.. controller에서 사용하더라도 지금과 같은 방식이 아닌 서비스 계층에서 던진 예외를 잡는 방식으로 사용해주세요ㅎㅎ

AlbertImKr replied:
개선 해보겠습니다.

janeljs commented on src/main/java/kr/codesqaud/cafe/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,37 @@
+    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);
+    public static final String ERROR_CODE = "errorCode";
+
+    @ExceptionHandler(CustomException.class)
```
지금은 CustomException으로 묶어서 처리하고 있지만 앞으로 적절하게 상속구조를 이용하여 예외를 계층화하거나 그룹핑해서 처리해도 좋을 것 같습니다ㅎㅎ

AlbertImKr replied:
넵
- Exception 계층화와 그룹핑를 한번 찾아보고 적용해보겠습니다. 🧑‍💻🧑‍💻🤦‍♂️

janeljs replied:
네ㅎㅎ 저도 아무것도 모를 때 적은 글이라 부끄럽기는 한데 아래 글도 참고해보세요ㅋㅋ
https://velog.io/@janeljs/codesquad-tech-talk-1

AlbertImKr replied:
감사합니다. 🙇🏻‍♂️🙇🏻‍♂️

janeljs reviewed (changes requested):
### Albert 안녕하세요~ 🤩 

![image](https://user-images.githubusercontent.com/68000537/230447003-0506e8be-a312-4c4c-b531-93bd0ca51ba4.png)

빠르게 PR 주셨네요~ 쉽지 않으셨을텐데 고생하셨습니다 💯 
고민하시는 지점과 추가로 고민해보셨으면 좋겠는 부분들 코멘트로 남겼으니 생각해보시고 답변 주세요 😄 
이번에는 빠르게 작업해주신 덕분에 머지 전까지 반영할 수 있는 시간이 있으실 것 같아 RC드립니다ㅎㅎ

AlbertImKr commented:
> 앨버트 AWS 배포하는 거 별로 안 어려우셨나요?? 나중에 하신다더니 뚝딱하셨네요.
> 
> 저는 새로운 개념에 대해 학습할 때 많이 헤매고 시간도 오래 걸리는데, 앨버트는 항상 뚝딱뚝딱 빨리 해내시는 것 보면 대단하다고 생각이 들어요. 특히 이번주가 저는 많이 헤매서 많이 못한 것 같은데 앨버트는 정말 많이 구현하셔서 대단합니다...

- AWS 는 깊게 공부하지는 못했고 블로그 따라 한번 해봤습니다. 일단 책을 구매했고 읽어 볼 예정입니다.  🧐
- 저는 아무래도 전에 이미 강의를 다 들어봤습니다. 기술을 조금 알고 있는 상태이여서 조금 빨리 할 수 있는 것 같습니다. 📚
- Spirng 처음 접한 Jinny 도 지금 잘하고 있다고 생각해요. 👍
- 감사합니다.

AlbertImKr commented:
> ### Albert 안녕하세요~ 🤩
> ![image](https://user-images.githubusercontent.com/68000537/230447003-0506e8be-a312-4c4c-b531-93bd0ca51ba4.png)
> 
> 빠르게 PR 주셨네요~ 쉽지 않으셨을텐데 고생하셨습니다 💯 고민하시는 지점과 추가로 고민해보셨으면 좋겠는 부분들 코멘트로 남겼으니 생각해보시고 답변 주세요 😄 이번에는 빠르게 작업해주신 덕분에 머지 전까지 반영할 수 있는 시간이 있으실 것 같아 RC드립니다ㅎㅎ

- 혼자서 많이 고민했고 답을 찾을 수 없었습니다.
- 남겨주신 코멘트로 보면서 찾아보고 해결해보겠습니다. 🤩🤩🤩

AlbertImKr committed:
feat: URL 수정

get:users/{userId} -> users/{userId}/profile
get:users/{userId}/update -> users/{userId}/profile/edit
put:users/{userId}/update -> users/{userId}/profile

AlbertImKr committed:
test: Test 수정

하드코딩된 실패 주소를 변경
URL 변경으로 인한 변경 URL 변경

AlbertImKr committed:
style: 네이밍 수정

UserService.containEmail -> containsEmail
getAllMembers -> getAllUsers

AlbertImKr committed:
refactor: MainController에서 PostRepository의존성 제거

AlbertImKr committed:
refactor: 필요 없는 에노테이션 제거

AlbertImKr committed:
test: @Nested 사용하여 테스트를 개선

AlbertImKr committed:
test: @Nested 사용하여 테스트를 개선

AlbertImKr committed:
test: 상수를 private로 설정

AlbertImKr committed:
style: 상위 RequestMapping 삭제

AlbertImKr committed:
style: userService.saveNewUser -> save로 네이밍 수정

AlbertImKr committed:
style: 상수처리

AlbertImKr committed:
feat:  DataBinder 기본 설정 initDirectFieldAccess() 로변경

AlbertImKr committed:
refactor: 비지니스 로직을 Service로 이동

AlbertImKr committed:
refactor: 접근자 변경

AlbertImKr committed:
fix: URL 잘못된 주소 수정

AlbertImKr committed:
refactor: Query 문 수정 및 NamedParameterJdbcTemplate 사용

AlbertImKr committed:
fix: 잘못 사용된 코드 isEmpty() -> isPresent로 수정

AlbertImKr committed:
test: 비지니스 로직의 변경으로 인한 테스트 수정

janeljs reviewed (approved):
Albert 리뷰 하나하나 확인하고 반영하시느라 정말 고생 많으셨습니다~ㅎㅎㅎ
이번 PR은 여기서 머지하고, 남겨주신 코멘트랑 수정사항은 차차 꼼꼼히 확인해볼게요 😄

AlbertImKr commented:
> Albert 리뷰 하나하나 확인하고 반영하시느라 정말 고생 많으셨습니다~ㅎㅎㅎ 이번 PR은 여기서 머지하고, 남겨주신 코멘트랑 수정사항은 차차 꼼꼼히 확인해볼게요 😄

- 저의 코드를 항상 꼼꼼히 체크해주시고 리뷰를 남겨주셔서 감사합니다.🙇🏻‍♂️🙇🏻‍♂️

janeljs commented on src/main/resources/templates/account/members.html:
```diff
@@ -43,16 +43,16 @@
                         <div class="d-grid gap-2 d-md-flex justify-content-md-end">
                             <button class="btn btn-primary" type="submit"
-                                    th:onclick="|location.href='@{/users/{id}/update(id = ${usersForm.id})}'|">수정
+                                    th:onclick="|location.href='@{/users/{id}/update(id = ${userForm.id})}'|">수정
```
url 수정이 안된 것 같아요~

AlbertImKr replied:
넵. 🤦‍♂️
